build/common.ps1

```diff
@@ -268,22 +270,74 @@ Function Install-DotnetCLIToILMergePack {
     & $cli.DotNetExe --info
 }
 
+Function Get-LatestVisualStudioRoot() {
+    # First try to use vswhere to find the latest version of Visual Studio.
+    if (Test-Path $VsWhereExe) {
+        $installationPath = & $VsWhereExe -latest -prerelease -property installationPath
+        Verbose-Log "Found Visual Studio at '$installationPath' using vswhere"
+
+        return $installationPath
+    }
+
+    $cachePath = "${Env:ProgramData}\Microsoft\VisualStudio\Packages"
+
+    # Fall back to reading JSON directly (not recommended).
+    'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\VisualStudio\Setup',
+    'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\VisualStudio\Setup',
+    'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\VisualStudio\Setup' | ForEach-Object {
+        if (Test-Path $_) {
+            $key = Get-ItemProperty $_
+
+            if ($key.CachePath) {
+                $cachePath = $key.CachePath
+                return
+            }
+        }
+    }
+
+    $instances = Get-ChildItem "$cachePath\_Instances" -Filter state.json -Recurse | Get-Content | ConvertFrom-Json
+    foreach ($instance in $instances) {
+        if (-not $maxInstance) {
+            $maxInstance = $instance
+        } elseif ([Version]$instance.installationVersion -gt [Version]$maxInstance.installationVersion) {
+            $maxInstance = $instance
+        } elseif (([Version]$instance.installationVersion -eq [Version]$maxInstance.installationVersion) -and ([DateTime]$instance.installDate -gt [DateTime]$maxInstance.installDate)) {
+            $maxInstance = $instance
+        }
+    }
+
+    if ($maxInstance) {
+        $installationPath = $maxInstance.installationPath
+        Verbose-Log "Found Visual Studio at '$installationPath' using machine configuration"
+
+        return $installationPath
+    }
+
+    Error-Log 'Cannot find an instance of Visual Studio 2017 or newer' -Fatal
+}
+
 Function Get-MSBuildRoot {
     param(
-        [ValidateSet(15)]
-        [int]$MSBuildVersion,
         [switch]$Default
     )
     # Willow install workaround
-    if (-not $Default -and $MSBuildVersion -eq 15 -and (Test-Path Env:\VS150COMNTOOLS)) {
-        # If VS "15" is installed get msbuild from VS install path
-        $MSBuildRoot = Join-Path $env:VS150COMNTOOLS ..\..\MSBuild
+    if (-not $Default) {
+        # Find version 15.0 or newer
+        if (Test-Path Env:\VS150COMNTOOLS) {
```

Comment:
This works for VS16?

Reply:
the else part, this will only be set for dev15. And this change adds an extension to that for dev16 with the else part.